File: no_arma/R/perf.R
# Code From:
# https://github.com/JuliaLang/Microbenchmarks/blob/master/perf.R


# nested for loop, trying to compute sum_{i=1}^{10000} (1/i^2), 500 times

## pi_sum ##
pisum = function() {
    t = 0.0
    for (j in 1:1000) {
        t = 0.0
        for (k in 1:10000) {
            t = t + 1.0/(k*k)
        }
    }
    return(t)
}

## pi_sum_vec ##
pisumvec = function() {
    r = 1:10000
    return(replicate(1000, sum(1/((r)^2)))[1])
}


## Fibonacci
fib = function(n) {
    if (n < 2) {
        return(n)
    } else {
        return(fib(n-1) + fib(n-2))
    }
}



assert = function(bool) {
    if (!bool) stop('Assertion failed')
}


perf <- function(){
    library(microbenchmark)
    filename <- "Rtimes.csv"
    conv_nanosecond_to_second <- 1e-9
    ans_pisum <- 1.644834071848065
    ans_fib <- 832040
    #ans_pisum <- 1.6449106278707863
    # should be ans_pisum <- pi^2 / 6
    # 1.6449340668482264061

    assert(abs(pisum()-ans_pisum) < 1e-12)
    # using nanoseconds, because seems to be buggy askign for seconds
    # unit = nanoseconds
    # options(microbenchmark.unit="ns")

    res <- microbenchmark(pisum(), times=21L, unit="ns")
    tmed_pisum <- median(res$time)

    # now convert to seconds
    tmed_pisum <- tmed_pisum * conv_nanosecond_to_second
    cat("median time is (pisum, for loop): ", tmed_pisum, " seconds \n", sep="")
    
    assert( fib(30) == ans_fib )
    res <- microbenchmark(fib(30), times=21L, unit="ns")
    tmed_fib <- median(res$time)

    # now convert to seconds
    tmed_fib <- tmed_fib * conv_nanosecond_to_second
    cat("median time is (fib, recursive): ", tmed_fib, " seconds \n", sep="")



    tmed <- c(tmed_pisum, tmed_fib)
    algoNames <- c("pisum", "fib")
    df <- data.frame(R=tmed)
    row.names(df) <- algoNames

    # write to table
    write.table(file=filename, x=df, sep=",", row.names=T, col.names=T, quote=F)
    cat("Writing to: ", filename, "\n", sep="")
}

perf()

File: no_arma/MATLAB/perf.m
% from https://github.com/JuliaLang/Microbenchmarks/blob/master/perf.m


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%  Main function. All the tests are run here.           %%
%%  The functions declarations can be found at the end.  %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function perf()
    filename = 'MATLABtimes.csv';
    lang = 'MATLAB';

    fileID = fopen(filename, 'w');
    fprintf(fileID, '%s\n', lang);
    fclose(fileID);
    %fprintf('%s\n', lang)

    %disp("starting pisum MATLAB")
    s = pisum(true);
    assert(abs(s-1.644834071848065) < 1e-9);
    timeit('pisum',@pisum, true)
    %disp("ending pisum MATLAB")


    %disp("starting fib")
    f = fib(30);
	assert(f == 832040)
	timeit('fib', @fib, 20)
    %disp("ending fib MATLAB")
end


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%  Functions declarations  %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function assert(bool)
   if ~bool
     error('Assertion failed')
   end
end

function timeit(name, func, varargin)
    filename = 'MATLABtimes.csv';

    lang = 'matlab';
    if exist('OCTAVE_VERSION') ~= 0
       lang = 'octave';
    end

    nexpt = 21;
    times = zeros(nexpt, 1);

    for i=1:nexpt
        tic(); func(varargin{:}); times(i) = toc();
    end

    times = sort(times);
    %#fprintf ('%s,%s,%.8f\n', lang, name, times(1)*1000);

    fileID = fopen(filename, 'a+');
    fprintf(fileID, '%s,%.8f\n', name, times(11)*1);
    fclose(fileID);
    %fprintf('%s,%.8f\n', name, times(11)*1000);
end

%% recursive fib %%

function f = fib(n)
    if n < 2
        f = n;
        return
    else
        f = fib(n-1) + fib(n-2);
    end
end


%% slow pi series %%

function sum = pisum(ignore)
    sum = 0.0;
    for j=1:1000
        sum = 0.0;
        for k=1:10000
            sum = sum + 1.0/(k*k);
        end
    end
end
